### DISCUSSION TITLE
Draw.io tools website URL-safe Base64 encoding + zip/unzip

### DISCUSSION THREAD

**Discussion by kriegaex** [jgraph/drawio#2544] [General]
* [x] I agree to follow the [Code of Conduct](https://github.com/jgraph/drawio/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
* [x] I have searched the issue tracker for a feature request that matches the one I want to file, without success.

**Is your feature request related to a problem? Please describe.**

I wanted to use https://jgraph.github.io/drawio-tools/tools/convert.html in order to perform the equivalent of the following shell command:

```shell
gzip -c file.txt | base64 -w0 | tr '/+' '_-' | tr -d '='
```

Unfortunately, there is no equivalent to the two `tr` commands available as an option in order to create a URL-safe Base64 string. Uuencoding it is not an option for me.

Furthermore, the "deflate" algorithm you use seems to be different from what I achieve with the `gzip` compression.

**Describe the solution you'd like**

I would like to be able to create a Base64 string equivalent to the above shell command. Finally, I want to be able to use it in a Groovy Web Console URL like [this one](https://gwc-experiment.appspot.com/?codez=H4sICAU63mEAA1JlZmVycmluZ1RvRGF0YVRhYmxlVmFyaWFibGVzVGVzdC5ncm9vdnkAbU9NT8JAEL3vr5gUDuDHChjRECQmevHiAYgXY8zQTsuG7W7d3QoJ6X93KE1TEw-bnffmvTczBcY7zAgSkj52Zf6Vow_kpA_csD_kUm338vt-NJ2O70ZTIVReWBfAFzbeSY0mk6stOkr-axQUq1TFGJQ1QsQavYclpeScMtnavmDANW40vaNTp9-vyQegQyCTePhjh6MAeGpGAaTKoAYDj_AgGCaUQpQShtIRfPQUX1C7Xk1Ch88Z9AIHv5X5htzVGSx5QYoGwzoXoOCNgjYQ9bvKfqvk-ngCz9p6njEYVpGojXTgLcOsroMr6czut-SoIds8mM9hMJbSDNtGnc1HdESXcCvlBC46XCtvprPhCNeLrusGJrDgV7G2EpX4BTXG2__XAQAA).

**Describe alternatives you've considered**

* shell commands
* an improved Groovy Web Console version creating the same URLs without having to resort to an external website

**Additional context**

See also https://github.com/groovy-console/groovy-web-console/issues/11.

**Comment by davidjgraph** [comment]
Offhand, I don't know. But gzip is not the same thing as deflate, though gzip generally uses deflate as part of the operation.

I would search for how to deflate using the command line:

https://stackoverflow.com/questions/3178566/how-to-deflate-with-a-command-line-tool-to-extract-a-git-object

Note the final operation is https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURI , not urlencode, I'm not aware of a built-in command line function that does that.

The operation sequence is:

encodeURI
deflate
base64 encode

**Comment by kriegaex** [comment]
If you follow the link to the other GitHub issue, you see the exact requirements and shell commands (see also above) to achieve what I need in this case. I think the description on the website I want to use is not entirely correct. It really zips and unzips the files, while Git really uses plain Deflate without any header information.

I am not saying that your free online tool has any bugs, hence the choice of the feature request template here. I am asking for additional options.

**Comment by davidjgraph** [comment]
Why not just leave the data uncompressed?

**Comment by kriegaex** [comment]
Why not just stick with the topic instead of starting a meta discussion? 😉

The resulting URLs are usually embedded as links in Stack Overflow answers or tutorials. The Groovy source code can get quite lengthy, sometimes containing multiple classes. Long source code means long, uncompressed URLs, more bandwidth consumption, longer page loading times. Compression can at least take some of the pressure off the system.

Counter question: Your web page is the only one I found offering compression in addition to Base64 encoding. Why did you implement that and not just offer uncompressed encoding? Maybe because it is more efficient or simply shorter, and that can be useful in some situations?

---

**Update:** If you close this as "won't fix" because you don't see any value in it or it is out of scope for the online tool, that is of course OK. I am not feeling entitled to anything here. I simply believe in continuous improvement and user feedback, being an agile coach in my daytime job. I thought my feature request would be a low-hanging fruit and easy enough to implement for you.
